pkg/chunk/openstack/swift_object_client.go

```diff
@@ -0,0 +1,207 @@
+	f.StringVar(&cfg.UserId, prefix+"swift.userid", "", "Openstack userid for the api.")
+	f.StringVar(&cfg.ApiKey, prefix+"swift.api-key", "", "Openstack api key.")
+	f.StringVar(&cfg.ApplicationCredentialId, prefix+"swift.app-credential-id", "", "Openstack Application Credential ID.")
+	f.StringVar(&cfg.ApplicationCredentialName, prefix+"swift.app-credential-name", "", "Openstack Application Credential Name.")
+	f.StringVar(&cfg.ApplicationCredentialSecret, prefix+"swift.app-credential-secret", "", "Openstack Application Credential Secret.")
+	f.StringVar(&cfg.AuthUrl, prefix+"swift.auth-url", "", "Openstack authentification URL.")
+	f.IntVar(&cfg.Retries, prefix+"swift.retries", 3, "Retries Openstack requests on failure.(default to 3)")
+	f.DurationVar(&cfg.Timeout, prefix+"swift.timeout", 60*time.Second, "The duration after which the requests to swift should be timed out. (default to 60s)")
+	f.DurationVar(&cfg.ConnectTimeout, prefix+"swift.connect-timeout", 10*time.Second, "The duration after which the connection to swift should be timed out. (default to 10s)")
+	f.StringVar(&cfg.Region, prefix+"swift.region", "", "Openstack Region to use eg LON, ORD - default is use first region (v2,v3 auth only)")
+	f.IntVar(&cfg.AuthVersion, prefix+"swift.auth-version", 0, "Openstack authentification version, set to 1, 2 or 3 or leave at 0 for autodetect.")
+	f.BoolVar(&cfg.Internal, prefix+"swift.internal", false, "Set this to true to use the the internal / service network.")
+	f.StringVar(&cfg.Tenant, prefix+"swift.tenant", "", "Openstack tenant name (v2,v3 auth only).")
+	f.StringVar(&cfg.TenantId, prefix+"swift.tenant-id", "", "Openstack tenant id (v2,v3 auth only).")
+	f.StringVar(&cfg.EndpointType, prefix+"swift.endpoint-type", "", "Endpoint type, possible values: public, admin, internal (v2,v3 auth only) (default is public URL unless Internal is set).")
+	f.StringVar(&cfg.TenantDomain, prefix+"swift.tenant-domain", "", "Name of the tenant's domain (v3 auth only), only needed if it differs from the user domain.")
+	f.StringVar(&cfg.TenantDomainId, prefix+"swift.tenant-domain-id", "", "Id of the tenant's domain (v3 auth only), only needed if it differs the from user domain.")
+	f.StringVar(&cfg.TrustId, prefix+"swift.trust-id", "", "Id of the trust (v3 auth only).")
+}
+
+// NewSwiftObjectClient makes a new chunk.Client that writes chunks to OpenStack Swift.
+func NewSwiftObjectClient(cfg SwitfConfig, delimiter string) (*SwiftObjectClient, error) {
+	util.WarnExperimentalUse("OpenStack Swift Storage")
+
+	if cfg.EndpointType != "" &&
+		cfg.EndpointType != string(swift.EndpointTypePublic) &&
+		cfg.EndpointType != string(swift.EndpointTypeAdmin) &&
+		cfg.EndpointType != string(swift.EndpointTypeInternal) {
+		return nil, fmt.Errorf(
+			"unexpected endpoint type %s must be one of %s, %s or %s",
+			cfg.EndpointType, swift.EndpointTypePublic, swift.EndpointTypeAdmin, swift.EndpointTypeInternal)
+	}
```

Comment:
As pattern, we're currently transitioning towards validating the config in a `cfg.Validate() error` function called by the parent config. I navigate back the (2) places where `SwiftConfig` is referenced and should be easy to call. As a reference you can see this:
https://github.com/cortexproject/cortex/blob/9451bc694c572212f3eefaa535427f30d2ac5468/pkg/chunk/storage/factory.go#L91-L100

Reply:
The rule store doesn't have a validate. I'll add it.

Reply:
Yes exactly. Just add the missing `Validate()` going back the parent, but should be trivial.